### PR TITLE
Adding attribute_constraints to AttributeByTemplate

### PR DESCRIPTION
**Pull request by maxhutch**
# Citrine Python PR

## Description 
These constraints apply to the attributes in the object template
that contains the target attribute.  This is client support for the backend feature added in https://github.com/CitrineInformatics/platform-backend/pull/885

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [x] I have added tests for 100% coverage
- [x] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.

### CODE REVIEW

**Comment by maxhutch** [comment]
Little example (not fully contained):
```
from taurus.entity.bounds.categorical_bounds import CategoricalBounds
defn_with_shape = defn_with_name.add_columns(
    description="Table with Name and Shape",
    variable=AttributeByTemplate(name="shape", headers=["Loaf", "Shape"],
                                 template=named_template("shape"),
                                 attribute_constraints=[(named_template("shape"), CategoricalBounds(categories=["vienna"]))]
                                ),
    columns=[MostLikelyCategoryColumn(data_source="shape")]
)
print(project.ara_definitions.preview(defn_with_shape, preview_roots))
>>> {'csv': 'Root~Name,Loaf~Shape~Mode,Root Citrine Identifier\nloaf,vienna,d2cce889-6230-4d3f-821a-f4858868a6f4\nloaf,,8132a2ba-bc59-462f-baf0-a3a555a6082d', 'warnings': []}
```

File: src/citrine/ara/variables.py [hunk 1]
```diff
@@ -97,12 +98,20 @@ class AttributeByTemplate(Serializable['AttributeByTemplate'], Variable):
         sequence of column headers
     template: LinkByUID
         attribute template that identifies the attribute to assign to the variable
+    attribute_constraints: list[(LinkByUID, Bounds)]
+        constraints, expressed as Bounds, on attributes, expressed with links, in the object that
```
Comment (kroenlein) [inline src/citrine/ara/variables.py]:
```suggestion
        constraints on object attributes in the target object that must be satisfied. Constraints
        are expressed as Bounds.  Attributes are expressed with links. The attribute that the 
        variable is being set to may be the target of a constraint as well.
```

File: src/citrine/ara/variables.py [hunk 2]
```diff
@@ -97,12 +98,20 @@ class AttributeByTemplate(Serializable['AttributeByTemplate'], Variable):
...
+        contains the target attribute.  The attribute that the variable is being set to may
```
Comment (kroenlein) [inline src/citrine/ara/variables.py]:
```suggestion
```

File: src/citrine/ara/variables.py [hunk 3]
```diff
@@ -97,12 +98,20 @@ class AttributeByTemplate(Serializable['AttributeByTemplate'], Variable):
...
+        be the target of a constraint as well.
```
Comment (kroenlein) [inline src/citrine/ara/variables.py]:
```suggestion
```

File: src/citrine/ara/variables.py [hunk 4]
```diff
@@ -131,13 +142,21 @@ class AttributeByTemplateAfterProcessTemplate(
         attribute template that identifies the attribute to assign to the variable
     process_template: LinkByUID
         process template that identifies the originating process
+    attribute_constraints: list[(LinkByUID, Bounds)]
+        constraints, expressed as Bounds, on attributes, expressed with links, in the object that
```
Comment (kroenlein) [inline src/citrine/ara/variables.py]:
```suggestion
        constraints on object attributes in the target object that must be satisfied. Constraints
        are expressed as Bounds.  Attributes are expressed with links. The attribute that the 
        variable is being set to may be the target of a constraint as well.
```

File: src/citrine/ara/variables.py [hunk 5]
```diff
@@ -131,13 +142,21 @@ class AttributeByTemplateAfterProcessTemplate(
...
+        contains the target attribute.  The attribute that the variable is being set to may
```
Comment (kroenlein) [inline src/citrine/ara/variables.py]:
```suggestion
```

File: src/citrine/ara/variables.py [hunk 6]
```diff
@@ -131,13 +142,21 @@ class AttributeByTemplateAfterProcessTemplate(
...
+        be the target of a constraint as well.
```
Comment (kroenlein) [inline src/citrine/ara/variables.py]:
```suggestion
```

File: src/citrine/ara/variables.py [hunk 7]
```diff
@@ -173,14 +194,21 @@ class AttributeByTemplateAndObjectTemplate(
         attribute template that identifies the attribute to assign to the variable
     object_template: LinkByUID
         template that identifies the associated object
-
+    attribute_constraints: list[(LinkByUID, Bounds)]
+        constraints, expressed as Bounds, on attributes, expressed with links, in the object that
```
Comment (kroenlein) [inline src/citrine/ara/variables.py]:
```suggestion
        constraints on object attributes in the target object that must be satisfied. Constraints
        are expressed as Bounds.  Attributes are expressed with links. The attribute that the
        variable is being set to may be the target of a constraint as well.
```

File: src/citrine/ara/variables.py [hunk 8]
```diff
@@ -173,14 +194,21 @@ class AttributeByTemplateAndObjectTemplate(
...
+        contains the target attribute.  The attribute that the variable is being set to may
```
Comment (kroenlein) [inline src/citrine/ara/variables.py]:
```suggestion
```

File: src/citrine/ara/variables.py [hunk 9]
```diff
@@ -173,14 +194,21 @@ class AttributeByTemplateAndObjectTemplate(
...
+        be the target of a constraint as well.
```
Comment (kroenlein) [inline src/citrine/ara/variables.py]:
```suggestion
```

**Pull request merged**